DRIVERS-716 Improved Bulk Write API (#1534)
mongodb/specifications: source/crud/bulk-write.md

corresponding namespace that defines the collection on which the operation should be performed.
Drivers SHOULD design this pairing in whichever way is most ergonomic for its language. For
example, drivers may:

- Include a required `namespace` field on each `WriteModel` variant.
- Accept a list of `(Namespace, WriteModel)` tuples for `models`.
- Define the following pair class:

```typescript
class NamespaceWriteModelPair {
    /**
     * The namespace on which to perform the write.
     */
    namespace: Namespace;

    /**
     * The write to perform.
     */
    model: WriteModel;
}
```

Drivers MUST throw an exception if the list provided for `models` is empty.

### Options

```typescript
class BulkWriteOptions {
    /**
     * Whether the operations in this bulk write should be executed in the order in which they were
     * specified. If false, writes will continue to be executed if an individual write fails. If
     * true, writes will stop executing if an individual write fails.
     *
     * Defaults to false.
     */
    ordered: Optional<Boolean>;

    /**
     * If true, allows the writes to opt out of document-level validation.
     *
     * Defaults to false.
     *

>      * This option is only sent if the caller explicitly provides a value.

Reviewer: FWIW, there doesn't appear to be any record of why the CRUD spec was modified to not sent `false` for `bypassDocumentValidation`. See: https://github.com/mongodb/specifications/pull/266#issuecomment-431491212

My last follow-up comment in [DRIVERS-450](https://jira.mongodb.org/browse/DRIVERS-450) (ticket created _after_ the spec change was merged) speculated that `bypassDocumentValidation: false` was unique in expressing a double-negative.

I could go either way here, but if we decide to keep the language as-is here, I'd suggest we also open a DRIVERS ticket to revise the CRUD spec and remove the ignorance of `false`. That can be a low priority change, as it's unlikely to affect an application and there's presently no test coverage anyway (see: [DRIVERS-1316](https://jira.mongodb.org/browse/DRIVERS-1316)).

>      * This option is only sent if the caller explicitly provides a value.

Reviewer 2: > I think we should leave this as-is.

Agreed. I am convinced to send what is set, and rely on the server to interpret the value. Suggest documenting in the Q&A section since it is a notable change with existing CRUD spec:

```md
### Why is `bypassDocumentValidation: false` sent?

`BulkWriteOptions.bypassDocumentValidation` is only sent if set (to either true or false).
`InsertOneOptions.bypassDocumentValidation` is only sent if set to true (not sent if false).

The rationale for only sending `InsertOneOptions.bypassDocumentValidation` when
true is not clear. DRIVERS-450 notes possible rationale.

Checking the values that users explicitly provide for options and omitting them
if they match the server default seems like unnecessary additional work for
drivers. It could also lead to issues in the unlikely event that the server
changes the default value for an option.
```

> speculated that `bypassDocumentValidation: false` was unique in expressing a double-negative.

I can see the double-negative being confusing. Though in the event `bypassDocumentValidation=false` is set and the user's intent is to opt-out of document validation, the result is doing an extra document validation (not harmful IMO).

>      * This option is only sent if the caller explicitly provides a value.

Author: Added a Q&A entry and filed DRIVERS-2865

     */
    bypassDocumentValidation: Optional<Boolean>;

    /**
     * A map of parameter names and values to apply to all operations within the bulk write. Value
     * must be constant or closed expressions that do not reference document fields. Parameters can
     * then be accessed as variables in an aggregate expression context (e.g. "$$var").
     *
     * This option is only sent if the caller explicitly provides a value.
     */
    let: Optional<Document>;

    /**
     * The write concern to use for this bulk write.
     *
     * NOT REQUIRED TO IMPLEMENT. Drivers MUST expose this option if retrieving a handle to a
     * client with a different write concern configured than that of the user's standard URI
     * options is nontrivial. Drivers MAY omit this option if they provide a way to retrieve a
     * lightweight handle to a client with a custom write concern configured, e.g. a
     * MongoClient.withWriteConcern() method.
     */
    writeConcern: Optional<WriteConcern>;

    /**
     * Whether detailed results for each successful operation should be included in the returned

>      * BulkWriteResult.

Reviewer: It may be helpful to note that this corresponds inversely to the value of `errorsOnly` in the `bulkWrite` command. Not necessary for other options since those have a direct counterpart.

     */
    deletedCount: Int64;

    /**
     * The results of each individual insert operation that was successfully performed.
     */
    insertResults: Map<Int64, InsertOneResult>;

    /**
     * The results of each individual update operation that was successfully performed.
     */
    updateResult: Map<Int64, UpdateResult>;

    /**
     * The results of each individual delete operation that was successfully performed.
     */
    deleteResults: Map<Int64, DeleteResult>;
}

class InsertOneResult {
    /**
     * The _id of the inserted document.
     */
    insertedId: Any;
}

class UpdateResult {
    /**
     * The number of documents that matched the filter.
     */
    matchedCount: Int64;

    /**
     * The number of documents that were modified.
     */
    modifiedCount: Int64;

    /**
     * The number of documents that were upserted.
     *
     * NOT REQUIRED TO IMPLEMENT. Drivers may choose not to provide this property so long as it is
     * always possible to discern whether an upsert took place.
     */
    upsertedCount: Int64;

    /**
     * The _id field of the upserted document if an upsert occurred.
     */

>     upsertedId: Optional<Any>;

Reviewer 2: > what is the API impact? Are you referring to the spec document itself?

I prefer leaving `upsertedCount` out of the spec unless needed. I expect `upsertedCount` can only be 0 or 1. `upsertedCount` is motivated by supporting upserting an `_id` with BSON NULL (seems rare IMO).

I interpret "NOT REQUIRED TO IMPLEMENT" to mean drivers are free to implement regardless of whether it is needed.

> This would require an envelope object in PHP, which would be clunkier than using a native `null` value and providing an `upsertedCount` field

Given the need, here is a revised suggestion:

```typescript
/**
 * The _id field of the upserted document if an upsert occurred.
 * It MUST be possible for a caller to discern whether an upsert took place, even if the `upsertedId` is a BSON NULL value.
 * If necessary, drivers MAY add a boolean `didUpsert` field.
*/
upsertedId: Optional<Any>;
```

>     upsertedId: Optional<Any>;

Reviewer: Fine to remove the explicit `upsertedCount` in exchange for some extra text under `upsertedId`.

I'd personally keep `upsertedCount` in PHP for consistency with existing APIs, but I'm not going to lobby for the suggested field name in this extra text.

implement this in one of the following ways:

- Expose the `hasVerboseResults` field in `BulkWriteResult` as defined above.
- Implement the `insertResults`, `upsertResults`, and `deleteResults` fields as optional types and
  document that they will be unset when `verboseResults` is false.
- Introduce separate `SummaryBulkWriteResult` and `VerboseBulkWriteResult` types.
  `VerboseBulkWriteResult` MUST have all of the required fields defined on `BulkWriteResult` above.
  `SummaryBulkWriteResult` MUST have all of the required fields defined on `BulkWriteResult` above
  except `insertResults`, `updateResults`, and `deleteResults`.

#### Individual Results

The `InsertOneResult`, `UpdateResult`, and `DeleteResult` classes are the same as or similar to
types of the same name defined in the [CRUD specification](crud.md). Drivers MUST redefine these
classes if their existing result classes deviate from the definitions in this specification (e.g.
if they contain acknowledgement information, which is not applicable for individual bulk write
operations). Drivers MAY reuse their existing types for these classes if they match the ones
defined here exactly.

> ### Exception

Reviewer: Oh, I missed that it referred to BulkWriteException already. Fine as-is.

     * the bulk write. This value may not be populated if the exception was thrown due to errors
     * occurring on individual writes.
     */
    error: Optional<Error>;

    /**
     * Write concern errors that occurred while executing the bulk write. This list may have
     * multiple items if more than one server command was required to execute the bulk write.
     */
    writeConcernErrors: WriteConcernError[];

    /**
     * Errors that occurred during the execution of individual write operations. This map will
     * contain at most one entry if the bulk write was ordered.
     */
    writeErrors: Map<Int64, WriteError>;

    /**
     * The results of any successful operations that were performed before the error was
     * encountered.
     */
    partialResult: Optional<BulkWriteResult>;
}
```

## Building a `bulkWrite` Command

The `bulkWrite` server command has the following format:

```json
{
    "bulkWrite": 1,
    "ops": <Array>,
    "nsInfo": <Array>,
    "errorsOnly": <Boolean>,
    "ordered": <Boolean>,
    "bypassDocumentValidation": <Boolean>,
    "comment": <Bson>,

>     "let": <Document>,

Reviewer: Elsewhere, the spec says that `errorsOnly` "MUST always be defined in the `bulkWrite` command". Is it actually optional on the server? If so, what does it default to? It'd be helpful to clarify that and other server-side defaults just so we have some accessible reference on that.

Speaking from experience with the other specs, this comes up from time to time and we can't always rely on the server manual to provide adequate command documentation.

>     "let": <Document>,

Reviewer: I'm on board with having drivers explicitly send default values for `ordered` and `errorsOnly`. Note that the CRUD spec has no language about only sending `ordered` when specified by the client, so I expect it is always passed over the wire.

>     "let": <Document>,

Reviewer 3: PyMongo always sends "ordered" on all insert/update/delete commands.

    ...additional operation-agnostic fields
}
```

> Drivers SHOULD use document sequences ([`OP_MSG`](../message/OP_MSG.rst) payload type 1) for the

Reviewer 2: ```suggestion
Drivers MUST use document sequences ([`OP_MSG`](../message/OP_MSG.rst) payload type 1) for the
```

Suggest preferring "MUST" over "SHOULD" to reduce possible variations in driver implementations. I expect it is easier to relax the spec later if needed.

Preferring "MUST" is also recommended in https://github.com/mongodb/specifications?tab=readme-ov-file#when-in-doubt-use-must-not-should-in-specs

> `ops` and `nsInfo` fields. Drivers MUST NOT use document sequences when auto-encryption is enabled.

Reviewer 2: > Are there any plans for to change that down the line?
Filed DRIVERS-2859 to request support of document sequences. I expect there would be a performance benefit.

> `ops` and `nsInfo` fields. Drivers MUST NOT use document sequences when auto-encryption is enabled.

Author: added a "Future Work" section with a link to that ticket

The `bulkWrite` command is executed on the "admin" database.

### Operations

The `ops` field is a list of write operation documents. The first entry in each document has
the name of the operation (i.e. "insert", "update", or "delete") as its key and the index in the
`nsInfo` array of the namespace on which the operation should be performed as its value. The

> documents have the following format:

Reviewer: Disregard my last comment. I wrote it before realizing it was an optimization and that an `ops` element referenced an index within `nsInfo`. I must have been thinking of a previous design from many months ago.

#### Insert

```json
{

>     "insert": <Int32>,

Reviewer: Is the server flexible about accepting an Int32 or Int64 here, or is Int32 absolutely required? I do realize how absurd an Int64 would be in this context.

    "document": <Document>
}
```

Drivers MUST add an `_id` field at the beginning of the insert document if one is not already

> present.

Reviewer: To leave nothing to the imagination, clarify that drivers should generate an [ObjectId](https://github.com/mongodb/specifications/blob/master/source/objectid.rst). FWIW, the CRUD spec isn't entirely clear on that.

> When a user executes a bulk write with an unacknowledged write concern, drivers SHOULD check the

Reviewer 2: ```suggestion
When a user executes a bulk write with an unacknowledged write concern, drivers MUST check the
```

size of the insert document to verify that it does not exceed `maxBsonObjectSize`. For acknowledged
bulk writes, drivers MAY rely on the server to return an error if the document exceeds
`maxBsonObjectSize`. The value for `maxBsonObjectSize` can be retrieved from the selected server's

> `hello` response.

Reviewer 2: Suggest adding a check for all documents in `ops` (not just inserts). This may inform a user when there is no chance of a write succeeding. Consider adding a separate section under `### Operations`:

```md
#### Unacknowledged size checks

When a user executes a bulk write with an unacknowledged write concern, drivers MUST check the size of documents in `ops` and error if there is no expected chance of success:

Check that the size of each insert document does not exceed `maxBsonObjectSize`. The value for `maxBsonObjectSize` can be retrieved from the selected server's `hello` response.

Check that the size of other documents does not exceed `maxBsonObjectSize` + 16384 (16 KiB). 16KiB is added to account for command overhead allowed in the server (refer: [SERVER-10643](https://jira.mongodb.org/browse/SERVER-10643)).

For acknowledged bulk writes, drivers MUST NOT do these checks and instead rely on the server for size checks.
```

Notably: C and Python drivers have different existing size checks for OP_MSG document sequences:

[The C driver](https://github.com/mongodb/mongo-c-driver/blob/f6547a375c8d0b9ae7eff898487835ad451ab3d0/src/libmongoc/src/mongoc/mongoc-write-command.c#L643-L647) does not check if the write is unacknowledged, and includes the 16k allowance:

```c
if (slen > max_bson_obj_size + BSON_OBJECT_ALLOWANCE) {
```

[The Python driver](https://github.com/mongodb/mongo-python-driver/blob/b041ca5f7c201f1b0ecb97c517a1ece8d775c00a/pymongo/message.py#L1340) checks if the write is unacknowledged, and does not include the 16k allowance:
```python
unacked_doc_too_large = not ack and (doc_length > max_bson_size)
```

Reviewer 2: Addressed by new `### Size Limits` section.

#### Update

```json
{
    "update": <Int32>,
    "filter": <Document>,

>     "updateMods": <Document | Array>,

Reviewer: This is a bizarre name choice since this can also be an update document, but I realize `update` would have conflicted with the first field. Either way, not your problem -- I just felt compelled to rant.

    "multi": <Boolean>,
    "upsert": Optional<Boolean>,
    "arrayFilters": Optional<Array>,
    "hint": Optional<String>
}
```

#### Delete

```json
{
    "delete": <Int32>,
    "filter": <Document>,
    "multi": <boolean>,
    "hint": Optional<String>,
    "collation": Optional<Document>
}
```

### Namespace Information

The `nsInfo` field is an array containing the namespaces on which the write operations should be

> performed. Drivers MUST NOT include duplicate namespaces in this list. The documents in the

Reviewer: Is the prohibition of duplicate namespaces enforced, or are we on the honor system here?

`nsInfo` array have the following format:

```json
{
    "ns": <String>
}
```

### `errorsOnly` and `verboseResults`

The `errorsOnly` field indicates whether the results cursor returned in the `bulkWrite` response
should contain only errors and omit individual results. If false, both individual results for
successful operations and errors will be returned.

This field corresponds to the `verboseResults` option defined on `BulkWriteOptions`. Its value
should be provided as `false` if `verboseResults` was set to `true`; otherwise, it should be
provided as `true`. This field MUST always be defined in the `bulkWrite` command regardless of
whether the user specified a value for `verboseResults`.

## Command Batching

Drivers MUST accept an arbitrary number of operations as input to the `MongoClient.bulkWrite` method.
Because the server imposes restrictions on the size of write operations, this means that a single
call to `MongoClient.bulkWrite` may require multiple `bulkWrite` commands to be sent to the server.
Drivers MUST split bulk writes into separate commands when the user's list of operations exceeds

> one or more of these maximums: `maxWriteBatchSize`, `maxBsonObjectSize`, and `maxMessageSizeBytes`.

Reviewer 2: Retracting suggestion. `maxBsonObjectSize` and `maxMessageSizeBytes` apply to both payload types.

Each of these values can be retrieved from the selected server's `hello` command response. Drivers
MUST merge results from multiple batches into a single `BulkWriteResult` or `BulkWriteException`
to return from `MongoClient.bulkWrite`.

### Number of Writes

> The `ops` array MUST NOT include more than `maxWriteBatchSize` operations.

Reviewer: Given the third prose test, I assume clients do not enforce this and instead rely on the server to do so. Should that be clarified here? I don't recall seeing it mentioned earlier.

### Total Message Size

#### Encrypted bulk writes

When auto-encryption is enabled, drivers MUST NOT provide the `ops` and `nsInfo` fields as document
sequences and MUST limit the size of the command according to the auto-encryption size limits
defined in the
[Client Side Encryption Specification](../client-side-encryption/client-side-encryption.rst).

#### Unencrypted bulk writes with document sequences

When `ops` and `nsInfo` are provided as document sequences, drivers MUST ensure that the total size
of the `OP_MSG` built for the `bulkWrite` command does not exceed `maxMessageSizeBytes`. Some
drivers may perform batch-splitting prior to constructing the full `OP_MSG` to be sent to the
server. In this case, drivers MAY use `maxMessageSizeBytes - 16,384` as the upper bound for the

> combined number of bytes in `ops` and `nsInfo`. 16KiB is subtracted to accommodate for the bytes in

Author: I'm not sure if other drivers will encounter a similar issue, but in Rust we build all of the operation-specific portions of a command prior to appending operation-agnostic fields and building an actual `OP_MSG`. This means that the total size of the message isn't known at the time of constructing `ops` and `nsInfo`. I lifted the 16 KiB number from the extra bytes the server allows in documents for overhead and it seems to be fine from my testing.

Reviewer: Sounds reasonable. Should you clarify that 16KiB is an approximation?

> the rest of the message.

Reviewer 2: Suggest adding a check for documents within the document sequence (not just inserts):

```md
Drivers MUST ensure each document within a document sequence does not exceed `maxBsonObjectSize` + 16384 if the write is unacknowledged. 16KiB is added to account for command overhead allowed in the server (refer: [SERVER-10643](https://jira.mongodb.org/browse/SERVER-10643)).
```

Notably: C and Python drivers have different existing size checks for OP_MSG document sequences:

[The C driver](https://github.com/mongodb/mongo-c-driver/blob/f6547a375c8d0b9ae7eff898487835ad451ab3d0/src/libmongoc/src/mongoc/mongoc-write-command.c#L643-L647) does not check if the write is unacknowledged, and includes the 16k allowance:

```c
if (slen > max_bson_obj_size + BSON_OBJECT_ALLOWANCE) {
```

[The Python driver](https://github.com/mongodb/mongo-python-driver/blob/b041ca5f7c201f1b0ecb97c517a1ece8d775c00a/pymongo/message.py#L1340) checks if the write is unacknowledged, and does not include the 16k allowance:
```python
unacked_doc_too_large = not ack and (doc_length > max_bson_size)
```

Reviewer 2: Retracting suggestion. Suggestion does not seem relevant to this section. This section describes batch splitting. Added suggestion to: https://github.com/mongodb/specifications/pull/1534/files#r1518880882

#### Unencrypted bulk writes without document sequences

When `ops` and `nsInfo` are provided within the `bulkWrite` command document and auto-encryption is
not enabled, drivers MUST ensure that the total size of the `bulkWrite` command document does not

> exceed `maxBsonObjectSize`.

Reviewer 2: Retracting suggestion. This section describes batch splitting. I expect the command overhead should not be included when batch splitting. Testing suggests sending a payload 0 sized to maxBsonObjectSize + 16384 results in an [error on mongos](https://github.com/kevinAlbs/c-bootstrap/blob/5cdd9262ccda033fcd3c0c3f5a737b73a4694529/tools/miniclient/test_opmsg_limits.py#L120-L128). I expect this is due to mongos internally adding fields to the sent document.

    "nInserted": <Int32>,
    "nUpserted": <Int32>,
    "nMatched": <Int32>,
    "nModified": <Int32>,
    "nDeleted": <Int32>,
    ...additional command-agnostic fields
}
```

If any operations were successful (i.e. `nErrors` is less than the number of operations that were
sent), drivers MUST record the summary count fields in a `BulkWriteResult` to be returned to the
user or embedded in a `BulkWriteException`. Drivers MUST NOT populate the `partialResult` field in
`BulkWriteException` if no operations were successful.

Drivers MUST attempt to consume the contents of the cursor returned in the server's `bulkWrite`
response before returning to the user. This is required regardless of whether the user requested
verbose or summary results, as the results cursor always contains any write errors that occurred.
If the cursor contains a nonzero cursor ID, drivers MUST perform `getMore`s until the cursor has
been exhausted. Drivers MUST use the same session used for the `bulkWrite` command for each
`getMore` call.

The documents in the results cursor have the following format:

```json
{
    "ok": <0 | 1>,
    "code": Optional<Int32>,
    "errmsg": Optional<String>,

>     "n": <Int32>,

Reviewer: Can we clarify exactly what these fields refer to? That was an omission of the CRUD spec and one reason why the deprecated bulk write spec remained relevant (as it summarized server responses).

Would `n` here only be relevant for insert and delete operations? I forget if there's relevance for update operations.

    "nMatched": <Int32>,
    "nModified": <Int32>,
    "upsertedId": Optional<ObjectId>

> }

Reviewer: Are there additional fields in the response not shown here? For instance, the CRUD spec refers to an `errInfo` field for write and write concern errors, which might be used to relay document validation errors.

Author: Update here: `errInfo` is the only other field expected to be in this response. Added it in with the other error fields.

```

Note that the responses do not contain information about the type of operation that was performed.
Drivers may need to maintain the user's list of write models to infer which type of result should

> be recorded.

Reviewer: How do drivers correlate cursor documents with original operations? Particularly, in an errorsOnly scenario, I assume the cursor would only report errorred operations. I don't see any field in the response that indicates the index of a write operation.

`errorsOnly` as `true`.

### `ordered`

The `ordered` field defines whether writes should be executed in the order in which they were
specified, and, if an error occurs, whether the server should halt execution of further writes. It
is optional and defaults to true on the server. Drivers MUST explicitly define `ordered` as `true`
in the `bulkWrite` command if a value is not specified in `BulkWriteOptions`. This is required to
avoid inconsistencies between server and driver behavior if the server default changes in the
future.

### Size Limits

The server reports a `maxBsonObjectSize` in its `hello` response. This value defines the maximum
size for documents that are inserted into the database. Documents that are sent to the server but
are not intended to be inserted into the database (e.g. command documents) have a size limit of
`maxBsonObjectSize + 16KiB`. When an acknowledged write concern is used, drivers MUST NOT perform
any checks related to these size limits and MUST rely on the server to raise an error if a limit is
exceeded. However, when an unacknowledged write concern is used, drivers MUST raise an error if one

> of the following limits is exceeded:

Author: Arguments I have for not checking these sizes for acknowledged writes:

- Relying on the server to return an error for oversized documents aligns with [this](https://github.com/mongodb/specifications/?tab=readme-ov-file#where-possible-depend-on-server-to-return-errors) driver specification mantra.
- Not checking sizes driver-side means the server can still proceed with other valid writes in the bulk write.
- If we have a hard cap of `maxMessageSizeBytes` for the total message size, acknowledged or unacknowledged, users will still be limited to what the server deems an acceptable number of bytes to be sent in one message.

Would additional language here to clarify that drivers should never send a message larger than `maxMessageSizeBytes` address your concerns about potential malicious input to this method? I ask because it seems like users do value their other writes being performed when an oversized document is specified, e.g. [this recent forum post](https://www.mongodb.com/community/forums/t/pymongo-insert-many-raising-operationfailure-instead-of-bulkwriteerror-on-oversized-documents/270467).

> of the following limits is exceeded:

Reviewer 2: > Not checking sizes driver-side means the server can still proceed with other valid writes in the bulk write.

Right, I forgot that point.

> Would additional language here to clarify

No need IMO. I think the batch splitting already addresses that messages will not exceed `maxMessageSizeBytes`.

- The size of a document to be inserted MUST NOT exceed `maxBsonObjectSize`. This applies to the
  `document` field of an `InsertOneModel` and the `replacement` field of a `ReplaceOneModel`.
- The size of an entry in the `ops` array MUST NOT exceed `maxBsonObjectSize + 16KiB`.

> - The size of the `bulkWrite` command document MUST NOT exceed `maxBsonObjectSize`. The overhead
>   bytes are omitted here intentionally to accommodate additional fields that mongos appends to the
>   command document.

Reviewer 2: ```suggestion
- The size of the `bulkWrite` command document MUST NOT exceed `maxBsonObjectSize + 16KiB`.
```

I prefer the driver only error when the the operation is certain not to succeed. Otherwise rely on the server to error.

When using unacknowledged writes, I expect there will be many cases where a write is likely to error, but a driver cannot reasonably check (e.g. server document validation). When in doubt, I rather err towards having the driver send what the user sets.

Author: that's reasonable, updated

new types also provides more clarity to users on the existing differences between the
collection-level and client-level bulk write APIs. For example, the `verboseResults` option is only
available for `MongoClient.bulkWrite`.

### Why are bulk write operation results returned in a cursor?

Returning results via a cursor rather than an array in the `bulkWrite` response allows full
individual results and errors to be returned without the risk of the response exceeding the
maximum BSON object size. Using a cursor also leaves open the opportunity to add `findAndModify` to
the list of supported write operations in the future.

### Why was the `verboseResults` option introduced, and why is its default `false`?

The `bulkWrite` command returns top-level summary result counts and, optionally, individual results
for each operation. Compiling the individual results server-side and consuming these results
driver-side is less performant than only recording the summary counts. We expect that most users
are not interested in the individual results of their operations and that most users will rely on
defaults, so `verboseResults` defaults to `false` to improve performance in the common case.

> ### Why are document sequences not permitted for encrypted bulk writes?
> 
> 

Reviewer 2: ```suggestion
```

Remove stray Q&A.

failing or a write concern error. Examples include network errors that occur when communicating
with the server, command errors (`{ "ok": 0 }`) returned from the server, client-side errors, and
errors that occur when attempting to perform a `getMore` to retrieve results from the server.

When a top-level error is encountered and individual results and/or errors have already been
observed, drivers MUST embed the top-level error within a `BulkWriteException` as the `error` field
to retain this information. Otherwise, drivers MAY throw an exception containing only the top-level
error.

Encountering a top-level error MUST halt execution of a bulk write for both ordered and unordered
bulk writes. This means that drivers MUST NOT attempt to retrieve more responses from the cursor or
execute any further `bulkWrite` batches and MUST immediately throw an exception.

### Write Concern Errors

Write concern errors are recorded in the `writeConcernErrors` field on `BulkWriteException`. When
a write concern error is encountered, it should not terminate execution of the bulk write for
either ordered or unordered bulk writes. However, drivers MUST throw an exception at the end of

> execution if any write concern errors were observed.

Reviewer 2: Suggest testing `writeConcernErrors` are collected across batches: https://github.com/mongodb/specifications/commit/44b7efe56099b8286fd80ebec7036cfff22af5f7

### Individual Write Errors

Individual write errors retrieved from the cursor are recorded in the `writeErrors` field on
`BulkWriteException`. If an individual write error is encountered during an ordered bulk write,
drivers MUST record the error in `writeErrors` and immediately throw the exception. Otherwise,

> drivers MUST continue to iterate the results cursor and execute any further `bulkWrite` batches.

Reviewer 2: Suggest testing `writeErrors` are collected across batches for unordered bulk writes: https://github.com/mongodb/specifications/commit/cf5743812843fdbaabee099cce85309033743c20

This revealed a bug in the C driver implementation (was not incrementing `txnNumber`).

Note that the responses do not contain information about the type of operation that was performed.
Drivers may need to maintain the user's list of write models to infer which type of result should
be recorded based on the value of `idx`.

### Handling Insert Results

Unlike the other result types, `InsertOneResult` contains an `insertedId` field that is generated
driver-side, either by recording the `_id` field present in the user's insert document or creating
and adding one. Drivers MUST only record these `insertedId`s in a `BulkWriteResult` when a
successful response for the insert operation (i.e. `{ "ok": 1, "n": 1 }`) is received in the
results cursor. This ensures that drivers only report an `insertedId` when it is confirmed that the
insert succeeded.

## Handling Errors

### Top-Level Errors

A top-level error is any error that occurs that is not the result of a single write operation
failing or a write concern error. Examples include network errors that occur when communicating
with the server, command errors (`{ "ok": 0 }`) returned from the server, client-side errors, and
errors that occur when attempting to perform a `getMore` to retrieve results from the server.

> 

Author: Added language and synced the test. We haven't done DRIVERS-2385 in Rust yet so I skipped it in my implementation, will assume it passes in C.

> 

Author: Added.

When a top-level error is encountered and individual results and/or errors have already been
observed, drivers MUST embed the top-level error within a `BulkWriteException` as the `error` field
to retain this information. Otherwise, drivers MAY throw an exception containing only the top-level
error.

Encountering a top-level error MUST halt execution of a bulk write for both ordered and unordered
bulk writes. This means that drivers MUST NOT attempt to retrieve more responses from the cursor or
execute any further `bulkWrite` batches and MUST immediately throw an exception. If the results
cursor has not been exhausted on the server when a top-level error occurs, drivers MUST send the

> `killCursors` command to attempt to close it.

Reviewer 2: ```suggestion
`killCursors` command to attempt to close it. The result from the `killCursors` command MAY be ignored.
```

I expect sending `killCursors` is a best-effort way to clean-up the cursor. If it fails, ignore the error. Suggested sentence is copied from [Find, getMore and killCursors commands](https://github.com/mongodb/specifications/blob/611b12ccbdd012dcd9ab2877a32200b3835c97af/source/find_getmore_killcursors_commands.rst#L446).

Author: updated

     */
    error: Optional<Error>;

    /**
     * Write concern errors that occurred while executing the bulk write. This list may have
     * multiple items if more than one server command was required to execute the bulk write.
     */
    writeConcernErrors: WriteConcernError[];

    /**
     * Errors that occurred during the execution of individual write operations. This map will
     * contain at most one entry if the bulk write was ordered.
     */
    writeErrors: Map<Int64, WriteError>;

    /**
     * The results of any successful operations that were performed before the error was
     * encountered.
     */
    partialResult: Optional<BulkWriteResult>;
}
```

## Building a `bulkWrite` Command

The `bulkWrite` server command has the following format:

```json
{
    "bulkWrite": 1,
    "ops": <Array>,
    "nsInfo": <Array>,
    "errorsOnly": Optional<Boolean>,
    "ordered": Optional<Boolean>,
    "bypassDocumentValidation": Optional<Boolean>,
    "comment": Optional<BSON value>,
    "let": Optional<Document>,
    ...additional operation-agnostic fields
}
```

Drivers MUST use document sequences ([`OP_MSG`](../message/OP_MSG.rst) payload type 1) for the

> `ops` and `nsInfo` fields. Drivers MUST NOT use document sequences when auto-encryption is enabled.

Author: `nsInfo` could exceed `maxBsonObjectSize` (~16MB) if

1. a user is doing `maxWriteBatchSize` operations,
2. each operation is on a different namespace, and
3. each namespace is close to the maximum length defined [here](https://www.mongodb.com/docs/manual/reference/limits/#mongodb-limit-Restriction-on-Collection-Names).

Just the namespaces would be 25MB in this scenario, and then there would be additional bytes to wrap each namespace in a document and then put those documents into an array.

Large `nsInfo` payloads could cause earlier batch splitting when they're used within the command document rather than as a document sequence. Batch splitting logic may also become more complex if adding one model to a bulk write `OP_MSG` requires size checks for both the command document and the overall message.

How much of a lift would it be to support multiple document sequences in the C driver? I'm inclined to leave in this requirement and expect drivers to update their `OP_MSG` implementations to support multiple document sequences, which is in line with the `OP_MSG` spec. FWIW, I made this change in the Rust driver https://github.com/mongodb/mongo-rust-driver/pull/1009/and it wasn't too much work.

> `ops` and `nsInfo` fields. Drivers MUST NOT use document sequences when auto-encryption is enabled.

Reviewer 2: > How much of a lift would it be to support multiple document sequences in the C driver?

It was not significant. Change has been applied to the C implementation.

Retracting suggestion.

> `ops` and `nsInfo` fields. Drivers MUST NOT use document sequences when auto-encryption is enabled.

Author: Sounds good, I'll add a Q&A entry for this.

future.

### Size Limits

The server reports a `maxBsonObjectSize` in its `hello` response. This value defines the maximum
size for documents that are inserted into the database. Documents that are sent to the server but
are not intended to be inserted into the database (e.g. command documents) have a size limit of
`maxBsonObjectSize + 16KiB`. When an acknowledged write concern is used, drivers MUST NOT perform
any checks related to these size limits and MUST rely on the server to raise an error if a limit is
exceeded. However, when an unacknowledged write concern is used, drivers MUST raise an error if one
of the following limits is exceeded:

- The size of a document to be inserted MUST NOT exceed `maxBsonObjectSize`. This applies to the
  `document` field of an `InsertOneModel` and the `replacement` field of a `ReplaceOneModel`.
- The size of an entry in the `ops` array MUST NOT exceed `maxBsonObjectSize + 16KiB`.
- The size of the `bulkWrite` command document MUST NOT exceed `maxBsonObjectSize + 16KiB`.

See [SERVER-10643](https://jira.mongodb.org/browse/SERVER-10643) for more details on these size
limits.

> ## Command Batching

Reviewer 4: Lets specify whether `nsInfo` contains only the namespaces actually used in the current batch, or all the namespaces passed to the `bulkWrite` method (as these can be different depending on the split points).  The latter may be simpler to implement, as it makes the logic for calculating the split points more straightforward (you can pre-calculate the size of `nsInfo` before you even execute the splitting logic).

Reviewer 4: @kevinAlbs I think C driver does the latter?

Author: I pushed a commit with more specific batch splitting logic plus a Q&A entry explaining the reasoning behind the design. This change should make implementation and testing of splitting up `nsInfo` simpler. @kevinAlbs, let me know if this proposed new logic makes sense for your implementation. I'm specifically curious whether differentiating between command-specific fields and command-agnostic fields is sensible. Approximating in this way will let us avoid a pretty major refactor in Rust. I will also add some tests around batch splitting on nsInfo that can rely on the 1000 bytes of overhead (i.e. the test proposed [here](https://github.com/mongodb/specifications/pull/1534#discussion_r1572908056)).

### Number of Writes

`maxWriteBatchSize` defines the total number of writes allowed in one command. Drivers MUST split a
bulk write into multiple commands if the user provides more than `maxWriteBatchSize` operations in
the argument for `models`.

### Total Message Size

#### Encrypted bulk writes

When auto-encryption is enabled, drivers MUST NOT provide the `ops` and `nsInfo` fields as document
sequences and MUST limit the size of each `bulkWrite` command according to the auto-encryption size
limits defined in the
[Client Side Encryption Specification](../client-side-encryption/client-side-encryption.rst).

#### Unencrypted bulk writes

When `ops` and `nsInfo` are provided as document sequences, drivers MUST ensure that the total size
of the `OP_MSG` built for each `bulkWrite` command does not exceed `maxMessageSizeBytes`. Some
drivers may perform batch-splitting prior to constructing the full `OP_MSG` to be sent to the
server. In this case, drivers MAY use `maxMessageSizeBytes - 16,384` as the upper bound for the
combined number of bytes in `ops` and `nsInfo`. 16KiB is subtracted as an approximate overhead
allowance to accommodate for the bytes in the rest of the message.

> 

Author: good idea, added

bulk write into multiple commands if the user provides more than `maxWriteBatchSize` operations in
the argument for `models`.

### Total Message Size

#### Encrypted bulk writes

When auto-encryption is enabled, drivers MUST NOT provide the `ops` and `nsInfo` fields as document
sequences and MUST limit the size of each `bulkWrite` command according to the auto-encryption size
limits defined in the
[Client Side Encryption Specification](../client-side-encryption/client-side-encryption.rst).

#### Unencrypted bulk writes

When `ops` and `nsInfo` are provided as document sequences, drivers MUST ensure that the total size
of the `OP_MSG` built for each `bulkWrite` command does not exceed `maxMessageSizeBytes`.

The upper bound for the size of an `OP_MSG` includes opcode-related bytes (e.g. the `OP_MSG`
header) and operation-agnostic command field bytes (e.g. `txnNumber`, `lsid`). Drivers MUST limit
the combined size of the `bulkWrite`-specific command fields, `ops` document sequence, and `nsInfo`
document sequence to `maxMessageSizeBytes - 1000` to account for this overhead. The following

> pseudocode demonstrates how to apply this limit in batch-splitting logic:

Author: Previous language for reference:

> Some drivers may perform batch-splitting prior to constructing the full `OP_MSG` to be sent to the server. In this case, drivers MAY use `maxMessageSizeBytes - 16,384` as the upper bound for the combined number of bytes in `ops` and `nsInfo`. 16KiB is subtracted as an approximate overhead allowance to accommodate for the bytes in the rest of the message.

The major difference here is that `bulkWrite` options are now excluded from the overhead allowance, which was problematic because _technically_ a `comment` or `let` document could be > 16,384 bytes and therefore cause the batch-splitting boundaries to be inaccurate.

method. Because the server imposes restrictions on the size of write operations, this means that a
single call to `MongoClient.bulkWrite` may require multiple `bulkWrite` commands to be sent to the
server. Drivers MUST split bulk writes into separate commands when the user's list of operations
exceeds one or more of these maximums: `maxWriteBatchSize`, `maxBsonObjectSize` (for `OP_MSG`
payload type 0), and `maxMessageSizeBytes` (for `OP_MSG` payload type 1). Each of these values can
be retrieved from the selected server's `hello` command response. Drivers MUST merge results from
multiple batches into a single `BulkWriteResult` or `BulkWriteException` to return from
`MongoClient.bulkWrite`.

When constructing the `nsInfo` array for a `bulkWrite` batch, drivers MUST only include the
namespaces that are referenced in the `ops` array for that batch.

### Number of Writes

`maxWriteBatchSize` defines the total number of writes allowed in one command. Drivers MUST split a
bulk write into multiple commands if the user provides more than `maxWriteBatchSize` operations in
the argument for `models`.

### Total Message Size

> #### Encrypted bulk writes

Reviewer 2: Out of caution, I suggest:

- Remove mention of supporting auto-encryption: https://github.com/mongodb/specifications/commit/ed483b5e9aee3f78d333cdbe9d9ea411b3d82770
- Return an error if auto-encryption is enabled: https://github.com/mongodb/specifications/commit/e39c82c4f68786d7dd29c0c1b4e1c342bf3cf069
- Add prose test to verify: https://github.com/mongodb/specifications/commit/62814a38b71d9edd796af9f5f0dacf487ead5e83

Rationale: This PR does not include tests for auto encryption. Tests are planned in DRIVERS-2888. As specified, a driver may implement `MongoClient.bulkWrite` from this PR and bypass use of libmongocrypt. That may result in a `MongoClient` configured with auto encryption sending a `bulkWrite` and ignoring auto encryption settings (may send plaintext when ciphertext was intended).

These changes may be soon-removed once DRIVERS-2888 is resolved. But I prefer to keep this PR self-contained. This may enable a driver to release `MongoClient.bulkWrite` without DRIVERS-2888.

## Specification

> > \[!NOTE\]

Reviewer: The escaping shouldn't be necessary here. See #1546
